### PR TITLE
Dev cad

### PR DESCRIPTION
**Pull request by janodw**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,8 @@ description = "A platform-agnostic driver for Semtech SX1276/77/78/79 based boar
 embedded-hal = "~0.2"
 bit_field = "~0.10"
 log = "0.4"
+bitflags = "1.2"
+rtt-target = { version = "0.3", features = ["cortex-m"] }
```
Comment (iamklob) [inline Cargo.toml]: Remove, no need to include specific output to driver lib.

File: src/lib.rs [hunk 2]
```diff
@@ -250,6 +270,22 @@ where
         Ok(())
     }
 
+    pub fn cad(
+        &mut self,
+        delay: &mut dyn DelayMs<u8>
+    ) -> Result<usize, Error<E, CS::Error, RESET::Error>>
+    {
+        rtt_target::rprintln!(" ----------- CAD start");
```
Comment (iamklob) [inline src/lib.rs]: Remove, see previous comment

**Review by iamklob — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (janodw) [reply Cargo.toml]: ok

File: src/lib.rs (hunk 2 above)
Comment (janodw) [reply src/lib.rs]: Ok, removed ..

File: src/lib.rs [hunk 3]
```diff
@@ -250,6 +270,32 @@ where
...
+        self.set_dio(FlagsDIO::DIO0_CAD_DONE, FlagsDIO::DIO0_GROUP)?;
+        self.set_dio(FlagsDIO::DIO1_CAD_DETECTED, FlagsDIO::DIO1_GROUP)?;
+        /* enable all interrupts */
+        self.write_register(Register::RegIrqFlagsMask.addr(), 0x00)?;
+        self.write_register(Register::RegIrqFlags.addr(), 0x05)?;
+        self.set_mode(RadioMode::CAD)?;
+        Ok(1)
+    }
+
+    pub fn sleep(
```
Comment (iamklob) [inline src/lib.rs:288]: This looks good. 2 questions:

- would it be useful to be able to tell if the module is currently in sleep mode (something like `radio.is_sleep()`?
- is there a way how to bring it back from sleep (possibly with the same register settings as there where before sleep)?

File: src/lib.rs [hunk 4]
```diff
@@ -291,8 +337,12 @@ where
         Ok(())
     }
 
-    pub fn set_dio0_tx_done(&mut self) -> Result<(), Error<E, CS::Error, RESET::Error>> {
```
Comment (iamklob) [inline src/lib.rs:294]: I would keep this method for backwards compatibility and use `set_dio` internally. Also I would mark it with [deprecated attribue](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-deprecated-attribute) possibly with a message directing user to use `set_dio` directly.

File: src/lib.rs (hunk 3 above)
Comment (janodw) [reply src/lib.rs:288]: > 

- would it be useful to be able to tell if the module is currently in sleep mode (something like radio.is_sleep()?

I don't see any reason for that. See my comment below ...

> 

- is there a way how to bring it back from sleep (possibly with the same register settings as there where before sleep)?

If I understand to documentation correctly, that changes of LoRa mode are through "mode changing" from Stand-By mode. And after then follows some specific type of configuration for this new mode (for Rx, Tx, or CAD). This means that a new mode configuration is always necessary.

File: src/lib.rs (hunk 4 above)
Comment (janodw) [reply src/lib.rs:294]: ok .. set_dio0_tx is back for backward compatibility

**Review by iamklob — APPROVED** [review]

**Pull request merged**